MayssaraOmar/assignment2/Assignment 2.cpp

```diff
@@ -0,0 +1,275 @@
+/* ASSUMPTIONS
+ * print_ancestors method: Binary Tree has unique keys
+ * common_ancestor method: if a node is an ancestor of the other node, return that node // that way there is always an answer
+ */
+#include <iostream>
+#include <stack>
+#include <utility>
+using namespace std;
+
+template<class T>
+class binary_tree;
+
+template<class T>
+class tree_node {
+	T data;
+	tree_node *left, *right;
+public:
+	tree_node(const T &data, tree_node* left, tree_node* right) :
+			data(data), left(left), right(right) {
+	}
+	friend class binary_tree<T> ;
+// the binary_tree class is declared as tree_node's friend which means that binary_tree can access the private members of tree_node
+};
+
+template<class T>
+class binary_tree {
```

Comment:
A note on design. It's not clear the separation of node and tree. Since several of the methods here take a node*, and those pointers could be in any tree, o no tree at all.
I think one some of the advantages you can have from a tree class (that you are not using) are:
* Finding a node by key.
* Ensuring all keys in a tree are different.